src/hub-connection.ts

```diff
@@ -1,78 +1,157 @@
+import { tap, map, filter, switchMap, skipUntil, debounceTime, take, delay } from "rxjs/operators";
 import { HubConnection as SignalRHubConnection } from "@aspnet/signalr-client";
 import { fromPromise } from "rxjs/observable/fromPromise";
 import { BehaviorSubject } from "rxjs/BehaviorSubject";
 import { Observable } from "rxjs/Observable";
+import { empty } from "rxjs/observable/empty";
+import { of } from "rxjs/observable/of";
 import { Observer } from "rxjs/Observer";
-import { tap } from "rxjs/operators";
 
 import { ConnectionState, ConnectionStatus, HubConnectionOptions } from "./hub-connection.model";
+import { buildQueryString } from "./utils/query-string";
+import { Dictionary } from "./utils/dictionary";
+import { emptyNext } from "./utils/rxjs";
 
 const connectedState: ConnectionState = { status: ConnectionStatus.connected };
+const connectionReadyState: ConnectionState = { status: ConnectionStatus.connectionReady };
 const disconnectedState: ConnectionState = { status: ConnectionStatus.disconnected };
 
 export class HubConnection<THub> {
 
 	get connectionState$() { return this._connectionState$.asObservable(); }
 
-	private _source: string;
-	private _hubConnection: SignalRHubConnection;
+	private source: string;
+	private hubConnection: SignalRHubConnection;
+	private hubConnectionOptions$: BehaviorSubject<HubConnectionOptions>;
 	private _connectionState$ = new BehaviorSubject<ConnectionState>(disconnectedState);
 
 	constructor(connectionOption: HubConnectionOptions) {
-		this._hubConnection = new SignalRHubConnection(connectionOption.endpointUri, connectionOption.options);
-		this._source = `[${connectionOption.key}] HubConnection ::`;
-	}
+		this.source = `[${connectionOption.key}] HubConnection ::`;
 
-	connect(): Observable<void> {
-		return fromPromise(this._hubConnection.start())
+		this.hubConnectionOptions$ = new BehaviorSubject<HubConnectionOptions>(connectionOption);
+
+		this.hubConnectionOptions$
```

Comment:
perhaps add a var for this, more so it get named e.g.

`reconnection$ = ...`
`reconnection$.subscribe();`